### DISCUSSION TITLE
How does Quarkus Kotlin/Java Interoperability work?

### DISCUSSION THREAD

**Discussion by cdennison** [quarkusio/quarkus#42163] [Q&A]
Hi all - also asked this here but didn't like @geoand's answer :)

https://stackoverflow.com/questions/77681674/how-does-quarkus-kotlin-java-interoperability-work

The question is can I mix regular Java and Kotlin in the same Quarkus instance OR does it have to be 100% Kotlin?

Thanks
Charlie
(Quarkus celebrity https://www.youtube.com/live/3-05a1Yyanw)

[a comment by an automated account is not shown]

**Reply by mschorsch** [reply]
> The question is can I mix regular Java and Kotlin in the same Quarkus instance OR does it have to be 100% Kotlin?

You can mix Java und Kotlin without problems.

**Reply by mschorsch** [reply]
Gradle: https://kotlinlang.org/docs/gradle-configure-project.html#kotlin-and-java-sources

Maven: https://kotlinlang.org/docs/maven.html#compile-kotlin-and-java-sources

**Reply by cdennison** [reply]
The issue I'm having isn't mixing Java and Kotlin - the issue is adding Kotlin to an Quarkus and dependency injection not working for me. Here's the full posting from SO:

I have a regular Quarkus Java application and I want to introduce some Kotlin classess via CDI and also the Event Bus but the following isn't working as I thought it would.

Case #1 - Event Bus

Kotlin Class

@ApplicationScoped
class GreetingServiceBus {
    @ConsumeEvent("KotlinTest")
    fun greeting(name: String) {
        System.out.println("hello $name");
    }
}
I call it from Resteasy class like this but greeting is never called (this works fine for Java Class with @ConsumeEvent("KotlinTest")

@Path("/api")
public class ApiExecute {
  @Inject
  EventBus eventBus;

  @POST
  @Path("/execute")
  public Response execute() {
    eventBus.send("KotlinTest", "Hello");

Case #2 - CDI

Kotlin Class

@ApplicationScoped
class GreetingService {
    fun greeting(name: String): String {
        return "hello $name"
    }
}
Resteasy Endpoint

@Path("/api")
public class ApiExecute {

  @Inject
  GreetingService greetingService;
Error

error: cannot find symbol GreetingService greetingService;

I saw something about using this with CDI "@field: Default" but I believe that's only available in Kotlin not Java?

Also tried constructor injection but same issue

  @POST
  @Path("/execute")
  @Inject
  public Response execute(GreetingService greetingService, String input) {

**Reply by stephan-strate** [reply]
How do you set up your maven-compiler-plugin and kotlin-compiler-plugin. Have you tried creating a minimal reproducer?

**Reply by geoand** [reply]
> How do you set up your maven-compiler-plugin and kotlin-compiler-plugin. Have you tried creating a minimal reproducer?

Exactly right, it would be super helpful to have such a reproducer

**Reply by mschorsch** [reply]
@cdennison I have created a simple Quarkus application 
[quarkus-java-kotlin-interoperability.zip](https://github.com/user-attachments/files/16404773/quarkus-java-kotlin-interoperability.zip)
in which Java and Kotlin classes are used together using dependency injection. I don't see any problems.

**Reply by cdennison** [reply]
I'm confirmed the reproducer works as expected and I added the event bus example confirming you can pub from java and sub from kotlin and vice versa.
[code-with-quarkus-java-kotlin-interop.zip](https://github.com/user-attachments/files/16417339/code-with-quarkus-java-kotlin-interop.zip)

I'm thinking about submitting it here? https://github.com/quarkusio/quarkus-quickstarts

I'm still trying to figure out why my original project is broken and will report back.

**Reply by geoand** [reply]
> I'm thinking about submitting it here? https://github.com/quarkusio/quarkus-quickstarts

That would be nice
